File: 0032-longest-valid-parentheses/0032-longest-valid-parentheses.cpp
class Solution {
public:
    int longestValidParentheses(string s) {
        int ans = 0;
        stack<int> st ;
        st.push(-1);
        for(int i=0;i<s.length();i++){
            int ch = s[i];
            if(ch == '(') st.push(i);
            else{
                st.pop();
                if(st.empty()) st.push(i);
                else{
                    int currentLength = i- st.top();
                    ans = max(ans,currentLength);
                }
                
            }
            
        }
        return ans;
    }
};

File: 0032-longest-valid-parentheses/0032-longest-valid-parentheses.java
class Solution {
    public int longestValidParentheses(String s) {
        int ans = 0;
        Stack<Integer> st = new Stack<>();
        st.push(-1);
        for(int i =0;i<s.length();i++){
            char ch = s.charAt(i);
            if(ch == '(') st.push(i);
            else{
                st.pop();
                if(st.isEmpty()){
                    st.push(i);
                }
                else{
                    int currentLength = i - st.peek();
                    ans = Math.max(ans, currentLength);
                }
            }
        }
        return ans;

    }
}